editor/components/post-permalink/index.js

```diff
@@ -1,79 +1,125 @@
-	}
-
-	componentWillUnmount() {
-		clearTimeout( this.dismissCopyConfirmation );
 	}
 
-	onCopy() {
-		this.setState( {
-			showCopyConfirmation: true,
-		} );
+	onVisibilityChange() {
+		const { isEditable, refreshPost } = this.props;
+		// If the user just returned after having clicked the "Change Permalinks" button,
+		// fetch a new copy of the post from the server, just in case they enabled permalinks.
+		if ( ! isEditable && 'visible' === document.visibilityState ) {
+			refreshPost();
+		}
 	}
 
-	onFinishCopy() {
-		this.setState( {
-			showCopyConfirmation: false,
-		} );
+	componentDidUpdate( prevProps, prevState ) {
+		// If we've just stopped editing the permalink, focus on the new permalink.
+		if ( prevState.isEditingPermalink && ! this.state.isEditingPermalink ) {
+			this.permalinkButton.focus();
+		}
 	}
 
 	render() {
-		const { isNew, link } = this.props;
-		if ( isNew || ! link ) {
+		const { isNew, previewLink, isEditable, samplePermalink } = this.props;
+		const { iconClass, isEditingPermalink } = this.state;
+
+		if ( isNew || ! previewLink ) {
 			return null;
 		}
 
 		return (
 			<div className="editor-post-permalink">
-				<Dashicon icon="admin-links" />
+				<Tooltip text={ __( 'Copy the permalink to your clipboard' ) }>
+					<ClipboardButton
+						className="editor-post-permalink__copy"
+						text={ samplePermalink }
+						onCopy={ () => this.setState( { iconClass: 'is-copied' } ) }
+					>
+						<Dashicon icon="admin-links" className={ iconClass } />
+					</ClipboardButton>
+				</Tooltip>
+
 				<span className="editor-post-permalink__label">{ __( 'Permalink:' ) }</span>
-				<Button className="editor-post-permalink__link" href={ link } target="_blank">
-					{ decodeURI( link ) }
-				</Button>
-				<ClipboardButton
-					className="button"
-					text={ link }
-					onCopy={ this.onCopy }
-					onFinishCopy={ this.onFinishCopy }
-				>
-					{ this.state.showCopyConfirmation ? __( 'Copied!' ) : __( 'Copy' ) }
-				</ClipboardButton>
+
+				{ ! isEditingPermalink &&
+					<Button
+						className="editor-post-permalink__link"
+						href={ previewLink }
+						target="_blank"
+						ref={ ( permalinkButton ) => this.permalinkButton = permalinkButton }
+					>
+						{ decodeURI( samplePermalink ) }
+						&lrm;
+					</Button>
+				}
+
+				{ isEditingPermalink &&
+					<PostPermalinkEditor
+						onSave={ () => this.setState( { isEditingPermalink: false } ) }
+					/>
+				}
+
+				{ isEditable && ! isEditingPermalink &&
+					<Button
+						className="editor-post-permalink__edit"
+						isLarge
+						onClick={ () => this.setState( { isEditingPermalink: true } ) }
+					>
+						{ __( 'Edit' ) }
+					</Button>
+				}
+
+				{ ! isEditable &&
+					<Button
+						className="editor-post-permalink__change"
+						isLarge
+						href={ getWPAdminURL( 'options-permalink.php' ) }
+						onClick={ window.addEventListener( 'visibilitychange', this.onVisibilityChange ) }
```

Comment:
This is not a callback. This will be binding to window whether or not the user clicks on the button. The binding needs to be within a function. We also need to unbind any manually-bound events in `componentWillUnmount`, otherwise they linger forever.

Reply:
👍🏻